src/sync.cpp

```diff
@@ -131,7 +131,7 @@ static void potential_deadlock_detected(const LockPair& mismatch, const LockStac
     throw std::logic_error("potential deadlock detected");
```

Comment:
I would suggest to unconditionally remove the entry from the stack here, after printing the stack and before `abort()`/`throw`. This cleanup is warranted from all code paths that call `EnterCritical()`.

```suggestion
    // Undo the insertion from push_lock(). We will not lock the mutex.
    s2.pop_back();
    throw std::logic_error("potential deadlock detected");
```

(this should be before the `if`, 4 lines above, but github does not allow me to put the suggestion there)